### PR TITLE
Rails 5.1

### PR DESCRIPTION
**Pull request by lihanli**
## Description of change
rails 5.1 upgrade

## Testing done
automated tests

## Testing planned
test on staging

#### Applies to all PRs

- [x] Appropriate logging
- [x] Swagger docs have been updated, if applicable
- [x] Provide link to originating GitHub issue, or connected to it via ZenHub
- [x] Does not contain any sensitive information (i.e. PII/credentials/internal URLs/etc., in logging, hardcoded, or in specs)
- [x] Provide which alerts would indicate a problem with this functionality (if applicable)

### CODE REVIEW

File: app/models/health_care_application.rb [hunk 1]
```diff
@@ -13,7 +13,9 @@ class HealthCareApplication < ApplicationRecord
   validates(:state, presence: true, inclusion: %w[success error failed pending])
   validates(:form_submission_id_string, :timestamp, presence: true, if: :success?)
 
-  after_save :send_failure_mail, if: proc { |hca| hca.state_changed? && hca.failed? && hca.parsed_form&.dig('email') }
+  after_save(:send_failure_mail, if: proc do |hca|
```
Comment (kfrz) [inline app/models/health_care_application.rb]: Can we extract this into a separate PR? Looks like it's not Rails 5.1 specific

File: spec/factories/form526_submissions.rb [hunk 2]
```diff
@@ -3,9 +3,9 @@
 FactoryBot.define do
   factory :form526_submission do
     user_uuid SecureRandom.uuid
-    saved_claim_id '123'
+    saved_claim { create(:va526ez) }
```
Comment (kfrz) [inline spec/factories/form526_submissions.rb]: Extract into separate PR?

File: spec/models/saved_claims/pension_spec.rb [hunk 3]
```diff
@@ -12,8 +12,8 @@
   context 'saved claims w/ attachments' do
     stub_virus_scan
 
-    let!(:attachment1) { FactoryBot.create(:pension_burial, saved_claim_id: nil) }
-    let!(:attachment2) { FactoryBot.create(:pension_burial, saved_claim_id: nil) }
+    let!(:attachment1) { FactoryBot.create(:pension_burial) }
```
Comment (kfrz) [inline spec/models/saved_claims/pension_spec.rb]: Extract into separate PR?

File: spec/models/user_preference_spec.rb [hunk 4]
```diff
@@ -10,26 +10,28 @@
 
   describe '.all_preferences_with_choices' do
     let(:user) { build(:user, :loa1) }
+    let(:account) { create(:account) }
```
Comment (kfrz) [inline spec/models/user_preference_spec.rb]: Extract into separate PR?

File: spec/factories/form526_submissions.rb [hunk 5]
```diff
@@ -3,9 +3,9 @@
...
     submitted_claim_id nil
-    auth_headers_json ''
+    auth_headers_json 'null'
```
Comment (lihanli) [inline spec/factories/form526_submissions.rb]: @kreek @SMLuthi in the new version of `attr_encrypted`, blank string for the `auth_headers_json` field violates the not-null db constraint, do we want to require a non-blank value for this field or remove the not-null db constrait?

File: app/models/health_care_application.rb (hunk 1 above)
Comment (lihanli) [reply app/models/health_care_application.rb]: `saved_change_to_attribute` does not work in rails 5.0

File: spec/factories/form526_submissions.rb (hunk 5 above)
Comment (kreek) [reply spec/factories/form526_submissions.rb]: @lihanli in the app flow auth headers should always be available so we could require a non-blank string.
